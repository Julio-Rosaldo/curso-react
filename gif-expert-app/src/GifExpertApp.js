import React, { Fragment, useEffect, useState } from 'react';
import { AddCategory } from './components/AddCategory';
import { ImageGrid } from './components/ImageGrid';

export const GifExpertApp = ( {defaultCategories = []} ) => {

    const [categories, setCategories] = useState(defaultCategories);

    function map() {
        // La funcion <array>.map regresa una funcion for each
        // que genera un array de los resultados
        return categories.map( (category) => {
            //return <li key={category}>{category}</li>
            return <ImageGrid key={category} category={category}/>
        })
    }

    // Cada vez que en el componente AddCategory se agregue un
    // elemento al array categories se renderiza todo ESTE
    // componente, provocando que ImageGrid se renderize nuevamente
    // lo que a su vez dispara la funcion useFetchGifs y tambien
    // renderiza nuevamente cada imagen de cada categoria

    // La renderizacion de los componentes no afecta a los valores
    // dinamicos que ya esten establecidos
    return (
        <Fragment>
            <h2>Gif Expert App</h2>
            {/* Se pasa la funcion setCategories */}
            <AddCategory setCategories={setCategories}/>
            <hr />
            {
                //<button onClick={handleAdd}>Agregar</button>
            }
            <ol>
                {map()}
            </ol>
        </Fragment>
    )
}