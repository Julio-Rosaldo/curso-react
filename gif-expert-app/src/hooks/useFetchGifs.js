import { useEffect, useState } from "react"
import { listGifs } from "../helpers/ListGifs";

// Hook a medida que orquestra el llamado a la consulta
// de imagenes
export const useFetchGifs = (category) => {

    const [state, setState] = useState({
        data: [],
        loading: true
    });

    /*
    setTimeout( () => {
        setState({
            data: [1, 2, 3, 4, 5],
            loading: false
        })
    }, 3000);
    */

    console.log("Hook " + category);

    // La funcion useEffect permite controlar la ejecucion de las
    // funciones despues de que se renderizan todos los componentes.

    // Esto implica la creacion implicida de un estado (useState)
    // que se mantiene en memoria cuyas referencias de acceso son
    // los valores que se especifican como segundo argumento de la
    // funcion. Si alguno de estos valores cambia, la funcion que
    // tenga como 1er argumento el useEffect sera ejecutada nuevamente
    // Si useEffect se ejecuta sin que alguno de estos valores
    // cambie, se regresara el estado en memoria que ya se tenia
    // previamente

    // En este caso, se evita llamar al API a menos que la categoria
    // cambie, lo cual es improbable para este caso
    useEffect(() => {
        listGifs(category).then( (images) => {
            setState({
                data: images,
                loading: false
            });
        } )
    }, [category])

    return state;
}