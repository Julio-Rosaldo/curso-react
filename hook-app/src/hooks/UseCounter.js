import React, { useState } from 'react'

export const UseCounter = (initialState = 1) => {
    const [count, setCounter] = useState(initialState);

    const increment = (factor = 1) => {
        setCounter(count + factor);
    }

    const decrement = (factor = 1) => {
        setCounter(count - factor);
    }

    function reset() {
        setCounter(initialState);
    }

    return {
        count,
        increment,
        decrement,
        reset
    }
}
