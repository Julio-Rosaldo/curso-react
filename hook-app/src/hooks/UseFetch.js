import React, { useEffect, useRef, useState } from 'react'

export const UseFetch = (url) => {

    const initialState = {
        data: null,
        loading: true,
        error: null
    };

    const isMounted = useRef(true);

    useEffect( () => {
        return () => {
            isMounted.current = false;
        }
    }, [])
  
    const [state, setState] = useState(initialState);

    useEffect( () => {

        setState(initialState);

        fetch(url)
        .then(response => response.json())
        .then(data => {

            setTimeout( () => {
                if (isMounted.current) {
                    setState({
                        loading: false,
                        error: null,
                        data: data
                    })
                } else {
                    console.log("setState no ejecutado");
                }
            }, 2000);
            
        });

    }, [url])

    return state;
}
