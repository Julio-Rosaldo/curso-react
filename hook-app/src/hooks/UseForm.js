import React, { useState } from 'react'

export const UseForm = (initialState = {}) => {

    const [values, setValues] = useState(initialState);

    const handleInputChange = (e) => {
        setValues({
            ...values,
            [e.target.name]: e.target.value
        })
    }

  return [values, handleInputChange];
}
