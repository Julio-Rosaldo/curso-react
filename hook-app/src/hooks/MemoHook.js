import React, { useMemo, useState } from 'react';
import { UseCounter } from '../hooks/UseCounter';
import "../components/Effects.css";
import { Small } from '../components/Small';

export const MemoHook = () => {

    const {count, increment} = UseCounter(5000);

    const [show, setShow] = useState(true);

    const procesoPesado = (iteraciones) => {
        for (let i = 0; i< iteraciones; i++) {
            console.log("Ciclo");
        }
        return iteraciones.toString() + ": iteraciones realizadas";
    }

    // Permite que una funcion solo se ejecute nuevamente si el parametro
    // indicado cambia de valor, eso sin importar si el componente se
    // renderiza nuevamente
    const memoProcesoPesado = useMemo( () => procesoPesado(count), [count]);

    console.log(show);

  return (
    <div>
        <h1>MemoHook</h1>
        <h1>Count: <Small value={count}></Small> </h1>
        <hr></hr>

        <p>{memoProcesoPesado}</p>

        <button className="btn btn-primary" 
        onClick={() => increment()}>+1</button>

        <p></p>
        <button className='btn btn-outline-primary ml-3'
        onClick={ () => {setShow(!show)} }>
            Show/Hide {JSON.stringify(show)}
        </button>
    </div>
  )
}
