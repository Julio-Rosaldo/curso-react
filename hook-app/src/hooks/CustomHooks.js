import React from 'react'
import { UseCounter } from './UseCounter';
import { UseFetch } from './UseFetch'

export const CustomHooks = () => {

    const {count, increment} = UseCounter(1);

    console.log(count);

    const url = "https://breakingbadapi.com/api/quotes/" + count.toString();

    const state = UseFetch(url);
    const {loading, error, data} = state;

    //const {author, quote} = !!data && data[0];
    let quote;
    if (data != null){
        quote = data[0];
    }
    console.log(quote);

  return (
    <div>
        <h1>Breaking Bad Quotes</h1>

{
    loading? (
    <div className="alert alert-info text-center">Loading...</div>
    )
    : 
    (
        <blockquote className="blockquote text-right">
        <p className="mb-0">{quote.quote}</p>
        <p></p>
        <footer className="blockquote-footer">{quote.author}</footer>
        </blockquote>
    )
}

<p>{count}</p>
        
        <button className="btn btn-primary" onClick={() => increment()}>Siguiente</button>
        
        
    </div>
  )
}
