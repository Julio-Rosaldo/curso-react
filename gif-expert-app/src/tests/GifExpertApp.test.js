import React from "react";
import "@testing-library/jest-dom";
import { GifExpertApp } from "../GifExpertApp";
import { shallow } from "enzyme";

import { useFetchGifs } from "../hooks/useFetchGifs";
jest.mock("../hooks/useFetchGifs");

describe("Pruebas en ", () => {

    test("Mostrar componente correctamente", () => {
        const wrapper = shallow(<GifExpertApp/>);
        expect(wrapper).toMatchSnapshot();
    })

    test("Mostrar lista de categorias", () => {
        const categories = ["Naruto", "One Piece"];
        const wrapper = shallow(<GifExpertApp defaultCategories={categories}/>);
        expect(wrapper).toMatchSnapshot();
    })

})