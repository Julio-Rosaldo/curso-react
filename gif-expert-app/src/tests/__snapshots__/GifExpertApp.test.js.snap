// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en  Mostrar componente correctamente 1`] = `
<Fragment>
  <h2>
    Gif Expert App
  </h2>
  <AddCategory
    setCategories={[Function]}
  />
  <hr />
  <ol />
</Fragment>
`;

exports[`Pruebas en  Mostrar lista de categorias 1`] = `
<Fragment>
  <h2>
    Gif Expert App
  </h2>
  <AddCategory
    setCategories={[Function]}
  />
  <hr />
  <ol>
    <ImageGrid
      category="Naruto"
      key="Naruto"
    />
    <ImageGrid
      category="One Piece"
      key="One Piece"
    />
  </ol>
</Fragment>
`;
